### PR TITLE
[FIX] tools: try decrypting PDF with empty password

### PR DESCRIPTION
**Pull request by backspac**
Steps:
- Install accounting
- On the dashboard, try to upload an "encrypted" PDF like this one https://launchpadlibrarian.net/24827090/DS-0157.pdf

Bug:
Traceback:
PyPDF2.utils.PdfReadError: file has not been decrypted

Explanation:
In 13.0, these errors were caught in a catch all `except` as shown here: https://github.com/odoo/odoo/blob/4e089041252c25cc197a30f9e285d82f7162d809/addons/account_facturx/models/account_move.py#L303-L328

From this SO comment: https://stackoverflow.com/questions/52047944/pdfbox-extracting-blanks-from-pdf-encrypted-with-no-password#comment91054285_52047944

> A PDF can be encrypted with two passwords: a user password and an owner password. When a PDF is encrypted with a user password, you can't open the document in a PDF viewer without entering that password. When a PDF is encrypted with an owner password only, everyone can open a PDF without that password, but some restrictions may be in place.

From time to time, we get a PDF encrypted with an owner password. The
content is still readable, but PyPDF2 fails because it thinks the PDF is
encrypted. With this fix, we try to unwrap the PDF by providing an empty
user password. This way, PyPDF2 thinks the content is now decrypted.

However, PyPDF2 only supports decrypting versions 1 and 2 of the
encryption implementation. If the version is different, we skip reading
the attachments and carry on to allow the user to upload the document.

opw:2375993

### CODE REVIEW

**Comment by backspac** [comment]
Ping @nim-odoo

**Review by xmo-odoo — CHANGES_REQUESTED** [review]
* unless `isEncrypted` can raise, it should be outside the `try...except`
* the two comments seem redundant, either is fine but both doesn't seem useful
* `_logger.exception` already embeds the entire formatted, it's unnecessary to format it again. Though in this case is formatting an entire traceback really useful? Maybe log as `warning`, and provide more information with respect to the PDF e.g. its name?
* talking about names, https://stackoverflow.com/a/26243460/8182118 mentions that PyPDF2 can't decrypt PDFs with spaces in the filename, not sure whether that applies here (I would assume not)
* wrt commit message, the subject should say what's being done / changed e.g. "try decrypting PDF with empty password"
* the commit message is also a bit... disjointed? the explanation doesn't really seem to explain much and I don't understand the relevance of the first line, surely if the error was discarded the user wouldn't see it? Or was the error discarded in 13 and isn't anymore in 14?

**Comment by xmo-odoo** [comment]
Also possibly useful background information, which would explain where / whence these odd PDFs-with-an-empty-password come from, might be useful to include (credited and linked) as background information in the commit message: https://stackoverflow.com/questions/52047944/pdfbox-extracting-blanks-from-pdf-encrypted-with-no-password#comment91054285_52047944

> A PDF can be encrypted with two passwords: a user password and an owner password. When a PDF is encrypted with a user password, you can't open the document in a PDF viewer without entering that password. When a PDF is encrypted with an owner password only, everyone can open a PDF without that password, but some restrictions may be in place.

**Comment by nim-odoo** [comment]
This might be worth noting as well:
> Note that this property, if true, will remain true even after the decrypt() method is called.

From [the documentation](https://pythonhosted.org/PyPDF2/PdfFileReader.html#PyPDF2.PdfFileReader.isEncrypted). So make sure we don't try to decrypt twice.

**Comment by xmo-odoo** [comment]
> This might be worth noting as well:
> 
> > Note that this property, if true, will remain true even after the decrypt() method is called.
> 
> From [the documentation](https://pythonhosted.org/PyPDF2/PdfFileReader.html#PyPDF2.PdfFileReader.isEncrypted). So make sure we don't try to decrypt twice.

Worth checking, but not sure it hurts:

> It checks the given password against the document’s user password and owner password, and then stores the resulting decryption key if either password is correct.

so decrypting twice would just pay for unnecessarily re-checking the password.

**Comment by nim-odoo** [comment]
As discussed elsewhere, we should also do the decrypt outside of the `getAttachment` method to respect the API of PyPDF. The goal of our override is to add the support for multiple attachments. Without our override, we would also have the decryption issue.

https://github.com/mstamy2/PyPDF2/pull/440/files

**Comment by odony** [comment]
Some extra things:

- Please be very explicit in both the commit msg and the code about the reason for making this "empty password" decryption step. We need to see that its purpose is to _unwrap_ PDF files that have an owner password set but not user password set, as the owner password is not necessary to simply read the PDF contents.
- Catching only `NotImplementedError` seems a bit unsafe - shouldn't we catch `Exception`, as we don't actually know what surprises PyPDF2 holds for decryption? Possibly keep the existing documentation of the NotImplementedError, but make it more defensive?
- Shouldn't the decryption happen when the reader is constructed, rather than as a side-effect of `getAttachments()`? (to be ready for other use cases)
- It would not hurt to have a test case for this. You could generate the encrypted file as part of the test, e.g. by extending [`test_odoo_pdf_file_writer()`](https://github.com/odoo/odoo/blob/master/odoo/addons/base/tests/test_pdf.py#L31-L44) and calling `pdf_writer.encrypt("", "foo")` at the end, then verifying that a new reader can access the attachments of the wrapped file? It's only a couple of lines.

Thanks!

**Comment by xmo-odoo** [comment]
> Catching only `NotImplementedError` seems a bit unsafe - shouldn't we catch `Exception`, as we don't actually know what surprises PyPDF2 holds for decryption? Possibly keep the existing documentation of the NotImplementedError, but make it more defensive?

Maybe we want to catch none of them and let the error bubble up in all cases, or convert a UserError to properly notify the user that something has gone wrong with the PDF parsing?

PyPDF2 documents that `NotImplementedError` is raised on unknown encryption methods, the user may want to know that we "can't open" the PDF in that case.

**Comment by backspac** [comment]
@xmo-odoo 
> talking about names, https://stackoverflow.com/a/26243460/8182118 mentions that PyPDF2 can't decrypt PDFs with spaces in the filename, not sure whether that applies here (I would assume not)

Indeed, it does not apply since we are giving it a `BytesIO`.

Also, calling `decrypt('')` multiple times does not change the outcome of the tests.

@odony 
> Shouldn't the decryption happen when the reader is constructed, rather than as a side-effect of `getAttachments()`? (to be ready for other use cases)

I think decrypting when it's needed is the best approach as they are already doing it in the lib: https://github.com/mstamy2/PyPDF2/blob/48319bee60683588ac012c5248b90e988a6c27c8/PyPDF2/pdf.py#L1208-L1216

**Comment by backspac** [comment]
@odony @xmo-odoo Is this OK to be merged?

**Review by odony — COMMENTED** [review]
thanks for the update! 2 minor things in the test could be fixed, the rest looks ready :+1:

@robodoo delegate+

**Comment by odony** [comment]
BTW I understand that you want to silence the `NotImplementedError` to allow users to upload unsupported files without being blocked. Yet from an API standpoint, it would make more sense to let `getAttachments()` raise when anything goes wrong. This way callers can distinguish an unsupported file from one with no attachments.
The callers are still free to silence those errors if *they* want to ignore it.

Basically that's what @xmo-odoo said in https://github.com/odoo/odoo/pull/62557#issuecomment-735736401
I have no preferences for NotImplementedError or UserError, but the former is closer to the native PyPDF2 behavior, and that's what we're trying to emulate/repair, so why not.

**Comment by backspac** [comment]
@odony Okay, thanks :)
I let the error bubble up out of the lib and caught it in the backend.

**Comment by backspac** [comment]
robodoo r+
